CONTRIBUTING.md

```diff
@@ -77,13 +77,29 @@ There are a handful of reasons the CI may have turned red. Try the following fix
 
 Every lint rule has an ID which can be used in lint directives (comments beginning with `#@ except:`) to prevent them from emitting diagnostics for portions of a WDL document. Rules "excepted" during the preamble (comments which are before the version statement) will be turned off for the entire document; otherwise, lint directives will shut off a rule while processing the children of whatever node it comes before, but only if that node is in the rule's `exceptable_nodes()` list. `exceptable_add()` will check all the ancestors of `element` for nodes that match the `exceptable_nodes()` list and see if they have a lint directive disabling the current rule; if so, the diagnostic will not be added to the validation output.
 
-#### Further reading
+### How should I decide which nodes to add to `exceptable_nodes()`?
+
+Returning `None` from `exceptable_nodes()` will enable lint directives to work anywhere in the document. This should be used sparingly, and is generally not the desired behavior. The most common case for returning `None` is if the lint rule pertains to "trivia" (whitespace and comments) which are permitted to appear almost anywhere in a WDL document.
+
+_Every_ rule that returns `Some(_)` for `exceptable_nodes()` should include `SyntaxKind::VersionStatementNode` as the first entry. This is to ensure the rule can be disabled for an entire document.
+
+The other `SyntaxKind` nodes that should be returned should be _intuitive_, _minimal_, and _comprehensive_.
+
+Intuitive: users shouldn't need to know anything about the internal CST or AST structure to determine where to add a lint directive. However, we can reasonably expect users to intuit the nested structure of a WDL document. There are top level items (workflow, task, and struct definitions), and then each of those top level items has a variety of sections contained within, which in turn may have further items contained within them. Excepting a rule somewhere "higher" in the nested structure will disable the rule for everything "beneath".
+
+Minimal: We want to have as few exceptable nodes as possible for each rule, while still enabling the most intuitive locations to be excepted. If two nodes would cover the same set of potential diagnostics, _always_ prefer the more specific node. e.g. for a rule relating to `input` sections, the `InputSectionNode`, `TaskDefinitionNode`, and `WorkflowDefinitionNode` would all cover the exact same set of diagnostics (as there can be at most one input section for each task or workflow definition); so we would prefer to have `InputSectionNode` returned by `exceptable_nodes()` rather than either `TaskDefinitionNode` or `WorkflowDefinitionNode`.
+
+Comprehensive: Try to cover as many _unique_ sets of potential diagnostics as possible with the nodes returned by `exceptable_nodes()`. An individual diagnostic should be able to targeted by a lint directive, and so should any unique group of diagnostics.
```

Comment:
> An individual diagnostic should be able to targeted by a lint directive, and so should any unique group of diagnostics.

I think this could be expanded. I wasn't clear on the first reading (and may not be now).

Reply:
Should I have an example here? The one coming to mind are the rules related to declarations, which can appear in a lot of places, and thus can have a decent amount of unique ways to "except" different groups of decls

Reply:
> Try to cover as many _unique_ sets of potential diagnostics as possible with the nodes returned by `exceptable_nodes()`.

Can you clarify what you mean by "unique sets"?  I think the second sentence is clear enough and an example would be noisy.

Reply:
e.g. https://github.com/stjude-rust-labs/wdl/blob/main/wdl-lint/src/rules/disallowed_declaration_name.rs#L57

All of the nodes returned here are potentially unique. The task and workflow definition nodes are a superset ofthe input node, output node, and private decls.